UnleashedRecomp/user/persistent_storage_manager.cpp

```diff
@@ -0,0 +1,126 @@
+#include "persistent_storage_manager.h"
+#include <install/installer.h>
+#include <os/logger.h>
+#include <user/paths.h>
+
+bool PersistentStorageManager::ShouldDisplayDLCMessage(bool setOffendingDLCFlag)
+{
+    auto result = false;
+
+    if (BinStatus != EBinStatus::Success)
+        return result;
+
+    static std::unordered_map<EDLCFlag, DLC> flags =
+    {
+        { EDLCFlag::ApotosAndShamar, DLC::ApotosShamar },
+        { EDLCFlag::Spagonia, DLC::Spagonia },
+        { EDLCFlag::Chunnan, DLC::Chunnan },
+        { EDLCFlag::Mazuri, DLC::Mazuri },
+        { EDLCFlag::Holoska, DLC::Holoska },
+        { EDLCFlag::EmpireCityAndAdabat, DLC::EmpireCityAdabat }
+    };
+
+    for (auto& pair : flags)
+    {
+        if (!Data.DLCFlags[(int)pair.first] && Installer::checkDLCInstall(GetGamePath(), pair.second))
+        {
+            if (setOffendingDLCFlag)
+                Data.DLCFlags[(int)pair.first] = true;
+
+            result = true;
+        }
+    }
+
+    return result;
+}
+
+bool PersistentStorageManager::LoadBinary()
+{
+    BinStatus = EBinStatus::Success;
+
+    auto dataPath = GetDataPath(true);
+
+    if (!std::filesystem::exists(dataPath))
+    {
+        // Try loading base persistent data as fallback.
+        dataPath = GetDataPath(false);
+
+        if (!std::filesystem::exists(dataPath))
+            return false;
+    }
+
+    std::error_code ec;
+    auto fileSize = std::filesystem::file_size(dataPath, ec);
+    auto dataSize = sizeof(PersistentData);
+
+    if (fileSize != dataSize)
+    {
+        BinStatus = EBinStatus::BadFileSize;
+        return false;
+    }
+
+    std::ifstream file(dataPath, std::ios::binary);
+
+    if (!file)
+    {
+        BinStatus = EBinStatus::IOError;
+        return false;
+    }
+
+    PersistentData data{};
+
+    file.read((char*)&data.Header.Signature, sizeof(data.Header.Signature));
+
+    if (!data.VerifySignature())
+    {
+        BinStatus = EBinStatus::BadSignature;
+        file.close();
+        return false;
+    }
+
+    file.read((char*)&data.Header.Version, sizeof(data.Header.Version));
+
+    if (!data.VerifyVersion())
+    {
+        BinStatus = EBinStatus::BadVersion;
+        file.close();
+        return false;
+    }
+
+    file.read((char*)&data.Header.HeaderSize, sizeof(data.Header.HeaderSize));
+
+    if (!data.VerifyHeader())
+    {
+        BinStatus = EBinStatus::BadHeader;
+        file.close();
+        return false;
+    }
```

Comment:
My idea with the header size wasn't to enforce a hard fail if the sizes don't match, it was to support incremental version updates, allowing older versions of recomp to skip the correct number of bytes in the header and read the data they know how to parse if the save file happens to be a newer version.

Buuut, this means older versions of recomp would also discard the newer data, which in hindsight kinda makes this safekeeping of the data it actually knows how to parse a bit pointless? What do you think?

If you want hard fail, then storing the header size isn't necessary.

Reply:
The header itself is unlikely to have any new additions as the data stored in this file isn't likely to be very important and future versions of the format would just add extra fields after the ones that are already there.

I'll remove the header size field for now and leave a couple reserved slots for future use if a new version of the format needs to expand upon it.

It'd be safer to create a new file should the user revert to a previous version of Unleashed Recompiled with a newer version of the format.